# Basic SQL Queries: Flights Database
© Stutern Graduate Accelerator

## Instructions to Students

This challenge is designed to determine how much you have learned so far and will test your knowledge on basic SQL queries.

#### Kindly write your answers in the answer section by double clicking the cell titled "YOUR ANSWER HERE"

Just incase you don't have the flights.db file, click [here](https://drive.google.com/drive/folders/1zoc3iqg3SMn2817C3ZAWeP2Bu-4FrvGS?usp=sharing) to download and ensure you upload it colab 

Do not add or remove cells in this notebook. Do not edit or remove the `%%sql` comment as it is required to run each cell.

### **FINAL NOTICE, DO NOT IGNORE**

* Make sure you select the right answers by running the right queries
* Once you are satisfied with your work, Kindly share a link to your notebook on **citrone**.
* Ensure access to notebook by changing general access from "Restricted" to "Anyone can view". This can be achieved by clicking the share button beside the comment button at the right hand side of the top menu/nav bar .

###**Thanks!**


**_Good Luck!_**

Below is an Entity Relationship Diagram (ERD) of the flights database:

<img src="https://github.com/Explore-AI/Pictures/blob/master/flights_ER_diagram.png?raw=true" width=40%/>

As can be seen from the ERD, the flights database consists of four tables:
- **flights**: all domestic flights in the USA in 2008
- **carriers**: lookup table for all the carriers
- **planes**: lookup table for all the planes 
- **airports**: lookup table for all the airports

For this challenge you will apply the use of Basic SQL queries to gain insight into the flights database. 

# DOUBLE CLICK HERE TO WRITE YOUR NAME

## NAME :  Aminat Owodunni
Name should appear as it is on citrone

## Loading the database

Before you start producing basic SQL queries, you need to prepare your SQL environment. You can do this by loading in the magic command `%load_ext sql` as we've seen in previous classes. Next you can go ahead and load in your database. To do this you will need to ensure you have downloaded the `flight.db` [here](https://drive.google.com/drive/folders/1zoc3iqg3SMn2817C3ZAWeP2Bu-4FrvGS?usp=sharing)  

In [ ]:
pip install ipython-sql

In [ ]:
pip install SQLAlchemy

In [ ]:
pip install pymysql

In [1]:
%load_ext sql

In [2]:
%%sql 

sqlite:///flights.db

In [3]:
%%time %%sql

UsageError: %%time is a cell magic, but the cell body is empty. Did you mean the line magic %time (single %)?


#Exploring the database 

Before we begin this challenge we can go ahead and take a look at the composition of each table in the flights database

In [4]:
%%sql 

PRAGMA table_info(flights);

 * sqlite:///flights.db
Done.


[(0, 'index', 'integer', 0, 'NULL', 0),
 (1, 'Date', 'longtext', 0, None, 0),
 (2, 'DayOfWeek', 'integer', 0, 'NULL', 0),
 (3, 'DepTime', 'double', 0, 'NULL', 0),
 (4, 'CRSDepTime', 'longtext', 0, None, 0),
 (5, 'ArrTime', 'double', 0, 'NULL', 0),
 (6, 'CRSArrTime', 'longtext', 0, None, 0),
 (7, 'UniqueCarrier', 'longtext', 0, None, 0),
 (8, 'FlightNum', 'longtext', 0, None, 0),
 (9, 'TailNum', 'longtext', 0, None, 0),
 (10, 'ActualElapsedTime', 'double', 0, 'NULL', 0),
 (11, 'CRSElapsedTime', 'longtext', 0, None, 0),
 (12, 'AirTime', 'longtext', 0, None, 0),
 (13, 'ArrDelay', 'double', 0, 'NULL', 0),
 (14, 'DepDelay', 'double', 0, 'NULL', 0),
 (15, 'Origin', 'longtext', 0, None, 0),
 (16, 'Dest', 'longtext', 0, None, 0),
 (17, 'Distance', 'integer', 0, 'NULL', 0),
 (18, 'TaxiIn', 'longtext', 0, None, 0),
 (19, 'TaxiOut', 'longtext', 0, None, 0),
 (20, 'Cancelled', 'longtext', 0, None, 0),
 (21, 'CancellationCode', 'longtext', 0, None, 0),
 (22, 'Diverted', 'longtext', 0, None, 0),
 (23, 'CarrierDelay', 'longtext', 0, None, 0),
 (24, 'WeatherDelay', 'longtext', 0, None, 0),
 (25, 'NASDelay', 'longtext', 0, None, 0),
 (26, 'SecurityDelay', 'longtext', 0, None, 0),
 (27, 'LateAircraftDelay', 'longtext', 0, None, 0)]

In [5]:
%%sql 

PRAGMA table_info(carriers);

 * sqlite:///flights.db
Done.


[(0, 'Code', 'varchar(10)', 0, 'NULL', 0),
 (1, 'Description', 'varchar(100)', 0, 'NULL', 0)]

In [6]:
%%sql 

PRAGMA table_info(airports);

 * sqlite:///flights.db
Done.


[(0, 'iata', 'varchar(10)', 0, 'NULL', 0),
 (1, 'airport', 'varchar(100)', 0, 'NULL', 0),
 (2, 'city', 'varchar(100)', 0, 'NULL', 0),
 (3, 'state', 'varchar(50)', 0, 'NULL', 0),
 (4, 'country', 'varchar(50)', 0, 'NULL', 0),
 (5, 'lat', 'float(24,0)', 0, 'NULL', 0),
 (6, 'long', 'float(24,0)', 0, 'NULL', 0)]

In [7]:
%%sql 

PRAGMA table_info(planes);

 * sqlite:///flights.db
Done.


[(0, 'tailnum', 'varchar(10)', 0, 'NULL', 0),
 (1, 'type', 'varchar(100)', 0, 'NULL', 0),
 (2, 'manufacturer', 'varchar(100)', 0, 'NULL', 0),
 (3, 'issue_date', 'varchar(100)', 0, 'NULL', 0),
 (4, 'model', 'varchar(50)', 0, 'NULL', 0),
 (5, 'status', 'varchar(100)', 0, 'NULL', 0),
 (6, 'aircraft_type', 'varchar(100)', 0, 'NULL', 0),
 (7, 'engine_type', 'varchar(100)', 0, 'NULL', 0),
 (8, 'year', 'longtext', 0, None, 0)]

## Questions on Basic SQL Queries 

Now that you have an idea of what information each table contains, you can use the given cell below each question to execute your basic SQL queries to find the correct output from the options provided for the multiple choice questions on form.

**Question 1**

How many different carriers are there in total in the database?

**Options:**
- 1 252
- 1491
- 20
- 37

In [8]:
%%sql
SELECT * FROM carriers
LIMIT 15;

 * sqlite:///flights.db
Done.


[('02Q', 'Titan Airways\r'),
 ('04Q', 'Tradewind Aviation\r'),
 ('05Q', 'Comlux Aviation, AG\r'),
 ('06Q', 'Master Top Linhas Aereas Ltd.\r'),
 ('07Q', 'Flair Airlines Ltd.\r'),
 ('09Q', 'Swift Air, LLC\r'),
 ('0BQ', 'DCA\r'),
 ('0CQ', 'ACM AIR CHARTER GmbH\r'),
 ('0FQ', 'Maine Aviation Aircraft Charter, LLC\r'),
 ('0GQ', 'Inter Island Airways, d/b/a Inter Island Air\r'),
 ('0HQ', 'Polar Airlines de Mexico d/b/a Nova Air\r'),
 ('0J', 'JetClub AG\r'),
 ('0JQ', 'Vision Airlines\r'),
 ('0KQ', 'Mokulele Flight Services, Inc.\r'),
 ('0LQ', 'Metropix UK, LLP.\r')]

In [9]:
%%sql
SELECT COUNT(Code AND Description) FROM carriers;

 * sqlite:///flights.db
Done.


[(1491,)]

### YOUR ANSWER HERE: 1491

**Question 2** 

How long was the longest delay before departure? 

**Options:**
- 1355 min
- 999 min
- 1099 min
- 588 min

In [10]:
%%sql
SELECT MAX(DepDelay) FROM flights

 * sqlite:///flights.db
Done.


[(1355.0,)]

### YOUR ANSWER HERE: 1355 min

**Question 3** 

How many flights departed on the 28th of January 2008?

**Options:**
- 540 908
- 20 149
- 19 495
- 18 056

In [11]:
%%sql
SELECT COUNT(*) FROM flights WHERE Date = '2008/1/28'

 * sqlite:///flights.db
Done.


[(20149,)]

### YOUR ANSWER HERE: 20149

**Question 4**
 
 What is the distance between Midway Airport (MDW) and Houston Airport (HOU)?
 
 **Options:**
 - 611 km
 - 972 km
 - 1 121 km
 - 937 km

In [12]:
%%sql
SELECT Origin, Dest, Distance FROM flights WHERE Origin = 'MDW' AND Dest = 'HOU'
LIMIT 1

 * sqlite:///flights.db
Done.


[('MDW', 'HOU', 937)]

### YOUR ANSWER HERE: 937km

**Question 5**

Which day of the week had the highest number of cancelled flights?(Note 1 = cancelled , 0 = not cancelled;)


**Options:**
- Monday(1)
- Wednesday(3) 
- Saturday(6) 
- Thursday (4) 

In [13]:
%%sql
SELECT DayOfWeek, COUNT(DayOfWeek) FROM flights WHERE Cancelled = 1  GROUP BY DayOfWeek

 * sqlite:///flights.db
Done.


[(1, 2617), (2, 2993), (3, 2645), (4, 3093), (5, 2049), (6, 1984), (7, 1927)]

### YOUR ANSWER HERE: Thursday(4)

**Question 6**

How many airports have the word "International" in their name?

**Options:**
- 110
- 124
- 8
- 2

In [14]:
%%sql
SELECT COUNT(airport) FROM airports WHERE airport LIKE '%International%'

 * sqlite:///flights.db
Done.


[(124,)]

### YOUR ANSWER HERE:124

**Question 7**

What is the most produced model for the manufacturer "BOEING"?

**Options:**
- 717-200
- 737-7H4
- 757-222
- 737-3H4

In [15]:
%%sql
SELECT model, SUM (DISTINCT model) FROM planes WHERE manufacturer = "BOEING" GROUP BY model

 * sqlite:///flights.db
Done.


[('717-200', 717.0),
 ('737-230', 737.0),
 ('737-236', 737.0),
 ('737-282', 737.0),
 ('737-282C', 737.0),
 ('737-2P6', 737.0),
 ('737-2X6C', 737.0),
 ('737-2Y5', 737.0),
 ('737-301', 737.0),
 ('737-317', 737.0),
 ('737-322', 737.0),
 ('737-33A', 737.0),
 ('737-3A4', 737.0),
 ('737-3B7', 737.0),
 ('737-3G7', 737.0),
 ('737-3H4', 737.0),
 ('737-3K2', 737.0),
 ('737-3L9', 737.0),
 ('737-3Q8', 737.0),
 ('737-3S3', 737.0),
 ('737-3T5', 737.0),
 ('737-3TO', 737.0),
 ('737-3Y0', 737.0),
 ('737-401', 737.0),
 ('737-490', 737.0),
 ('737-4B7', 737.0),
 ('737-4Q8', 737.0),
 ('737-4S3', 737.0),
 ('737-522', 737.0),
 ('737-524', 737.0),
 ('737-5H4', 737.0),
 ('737-705', 737.0),
 ('737-724', 737.0),
 ('737-73A', 737.0),
 ('737-76N', 737.0),
 ('737-76Q', 737.0),
 ('737-790', 737.0),
 ('737-7AD', 737.0),
 ('737-7BD', 737.0),
 ('737-7H4', 737.0),
 ('737-7Q8', 737.0),
 ('737-824', 737.0),
 ('737-832', 737.0),
 ('737-890', 737.0),
 ('737-8FH', 737.0),
 ('737-924', 737.0),
 ('737-924ER', 737.0),
 ('737-990', 737.0),
 ('747-2B5F', 747.0),
 ('747-422', 747.0),
 ('747-451', 747.0),
 ('757-212', 757.0),
 ('757-222', 757.0),
 ('757-223', 757.0),
 ('757-224', 757.0),
 ('757-225', 757.0),
 ('757-231', 757.0),
 ('757-232', 757.0),
 ('757-23N', 757.0),
 ('757-251', 757.0),
 ('757-26D', 757.0),
 ('757-2B7', 757.0),
 ('757-2G7', 757.0),
 ('757-2Q8', 757.0),
 ('757-2S7', 757.0),
 ('757-324', 757.0),
 ('757-33N', 757.0),
 ('757-351', 757.0),
 ('767-201', 767.0),
 ('767-223', 767.0),
 ('767-224', 767.0),
 ('767-2B7', 767.0),
 ('767-322', 767.0),
 ('767-323', 767.0),
 ('767-324', 767.0),
 ('767-332', 767.0),
 ('767-33A', 767.0),
 ('767-3CB', 767.0),
 ('767-3G5', 767.0),
 ('767-3P6', 767.0),
 ('767-424ER', 767.0),
 ('767-432ER', 767.0),
 ('777-222', 777.0),
 ('777-224', 777.0),
 ('777-232', 777.0),
 ('777-232LR', 777.0)]

### YOUR ANSWER HERE: 757-222

**Question 8**

How many planes landed at Los Angeles International Airport? 

**Options:**
- 215 608
- 18 964
- 41 258
- 39 422

In [16]:
%%sql
SELECT airport, COUNT(Dest) FROM airports
LEFT JOIN flights ON airports.iata = flights.Dest
WHERE airport = 'Los Angeles International'

 * sqlite:///flights.db
Done.


[('Los Angeles International', 18964)]

### YOUR ANSWER HERE: 18964